File: top/1001. Battle Over Cities - Hard Version (35).cpp
#include<iostream>
#include<algorithm>
using namespace std;
const int inf=0x3f3f3f3f;
const int maxn=1e5;
int fa[maxn],cost[maxn];
struct node{
	int from,to;
	int w,s;
	void read(){cin>>from>>to>>w>>s;}
	bool operator < (const node &a) const{
		if(s==a.s)return w<a.w;
		return s>a.s;
	}
}v[maxn];
int find(int x){
	int a=x;
	while(x!=fa[x])x=fa[x];
	while(a!=fa[x]){
		int tmp=a;
		a=fa[a];
		fa[tmp]=x;
	}
	return x;
}
int main(){
	ios::sync_with_stdio(false);
	int n,m,res=0;
	cin>>n>>m;
	for(int i=0;i<m;++i)v[i].read();
	sort(v,v+m);
	for(int i=1;i<=n;++i){
		for(int j=1;j<=n;++j)fa[j]=j;
		int cnt=n-2;
		cost[i]=0;
		for(int j=0;j<m;++j){
			if(v[j].from==i||v[j].to==i)continue;
			int fx=find(v[j].from),fy=find(v[j].to);
			if(fx==fy)continue;
			--cnt;
			if(!v[j].s)cost[i]+=v[j].w;
			fa[fx]=fy;
		}
		if(cnt>0)cost[i]=inf;
		res=max(cost[i],res);
	}
	if(!res){
		cout<<0;
		return 0;
	}
	bool flag=false;
	for(int i=1;i<=n;++i)
		if(res==cost[i]){
			if(flag)cout<<" ";
			flag=true;
			cout<<i;
		}
	return 0;
}

File: top/java/1001. Battle Over Cities - Hard Version (35).java
import java.util.Arrays;
import java.util.Scanner;

public class Main {
	public static final int maxn=100010;
	public static final int inf=0x3f3f3f3f;
	public static int fa[]=new int[maxn];
	public static int cost[]=new int[maxn];
	public static void main(String[] args) {
		Scanner in=new Scanner(System.in);
		int n=in.nextInt(),m=in.nextInt();
		node v[]=new node[m];
		for(int i=0;i<m;++i) {
			v[i]=new node(in.nextInt(),in.nextInt(),in.nextInt(),in.nextInt());
		}
		Arrays.sort(v);
		int res=0;
		for(int i=1;i<=n;++i) {
			for(int j=1;j<=n;++j)fa[j]=j;
			int cnt=n-2;
			cost[i]=0;
			for(int j=0;j<m;++j) {
				if(v[j].a==i||v[j].b==i)continue;
				int fx=find(v[j].a),fy=find(v[j].b);
				if(fx==fy)continue;
				--cnt;
				fa[fx]=fy;
				if(v[j].s==0)cost[i]+=v[j].c;
			}
			if(cnt>0)cost[i]=inf;
			res=Math.max(res, cost[i]);
		}
		if(res!=0) {
			boolean first=true;
			for(int i=1;i<=n;++i)
				if(cost[i]==res) {
					if(!first)System.out.print(" ");
					first=false;
					System.out.print(i);
				}
		}else System.out.print(res);
	}

	public static int find(int x) {
		int a=x;
		while(x!=fa[x])x=fa[x];
		while(a!=fa[a]) {
			int tmp=a;
			a=fa[a];
			fa[tmp]=x;
		}
		return x;
	}
}

class node implements Comparable<node>{
	int a,b;
	int c,s;

	node(int aa,int bb,int cc,int ss){
		this.a=aa;
		this.b=bb;
		this.c=cc;
		this.s=ss;
	}

	@Override
	public int compareTo(node n1) {
		if(n1.s==s) {
			if(c<n1.c)return -1;
			else return 1;
		}
		if(n1.s<s)return -1;
		return 1;
	}
}